From 60e8382e5193a49aa1723749dcad855b9d1e0829 Mon Sep 17 00:00:00 2001
From: Jiafei Pan <Jiafei.Pan@nxp.com>
Date: Thu, 26 Oct 2023 15:15:13 +0800
Subject: [PATCH 08/23] mailbox: gen-sw-mailbox: fix handle sequence

Change mailbox state machine only after the message have already
be handled.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>
---
 drivers/mailbox/generic-software-mailbox.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/mailbox/generic-software-mailbox.c b/drivers/mailbox/generic-software-mailbox.c
index 459770c99d1e..2961463bdddf 100644
--- a/drivers/mailbox/generic-software-mailbox.c
+++ b/drivers/mailbox/generic-software-mailbox.c
@@ -128,11 +128,11 @@ static irqreturn_t sw_mbox_interrupt(int irq, void *dev_id)
 		rx_status = readl(&mbox->base->rx_status[i]);
 		if (rx_status == S_BUSY) {
 			rx_ch = readl(&mbox->base->rx_ch[i]);
+			mbox_chan_received_data(&mbox->chan[i + RX_CHAN_SHFT],
+						(void *)&rx_ch);
 			writel(S_DONE, &mbox->base->rx_status[i]);
 			irq_set_irqchip_state(mbox->remote_irq,
 					      IRQCHIP_STATE_PENDING, true);
-			mbox_chan_received_data(&mbox->chan[i + RX_CHAN_SHFT],
-						(void *)&rx_ch);
 			ret = IRQ_HANDLED;
 		}
 	}
@@ -140,11 +140,11 @@ static irqreturn_t sw_mbox_interrupt(int irq, void *dev_id)
 	for (i = 0; i < MBOX_RXDB_CHAN; i++) {
 		rxdb_status = readl(&mbox->base->rxdb_status[i]);
 		if (rxdb_status == S_BUSY) {
+			mbox_chan_received_data(&mbox->chan[i + RXDB_CHAN_SHFT],
+						NULL);
 			writel(S_DONE, &mbox->base->rxdb_status[i]);
 			irq_set_irqchip_state(mbox->remote_irq,
 					      IRQCHIP_STATE_PENDING, true);
-			mbox_chan_received_data(&mbox->chan[i + RXDB_CHAN_SHFT],
-						NULL);
 			ret = IRQ_HANDLED;
 		}
 	}
-- 
2.34.1

